xarray/core/variable.py

```diff
@@ -463,77 +471,67 @@ def _nonzero(self):
         return tuple(Variable((dim), nz) for nz, dim
                      in zip(nonzeros, self.dims))
 
-    def _broadcast_indexes_advanced(self, key):
-        if isinstance(self._data, dask_array_type):
-            # dask only supports a very restricted form of advanced indexing
-            return self._broadcast_indexes_dask_pointwise(key)
+    def _broadcast_indexes_vectorized(self, key):
 
         variables = []
+        out_dims_set = OrderedSet()
         for dim, value in zip(self.dims, key):
             if isinstance(value, slice):
-                value = np.arange(*value.indices(self.sizes[dim]))
-
-            try:
-                variable = as_variable(value, name=dim)
-            except MissingDimensionsError:  # change to better exception
-                raise IndexError("Unlabelled multi-dimensional array "
-                                 "cannot be used for indexing.")
-
-            if variable.dtype.kind == 'b':  # boolean indexing case
-                if variable.ndim > 1:
-                    raise IndexError("{}-dimensional boolean indexing is "
-                                     "not supported. ".format(variable.ndim))
-                variables.extend(variable._nonzero())
+                out_dims_set.add(dim)
             else:
+                try:
+                    variable = as_variable(value, name=dim)
+                except MissingDimensionsError:  # change to better exception
+                    raise IndexError("Unlabelled multi-dimensional array "
+                                     "cannot be used for indexing.")
+
+                if variable.dtype.kind == 'b':  # boolean indexing case
+                    if variable.ndim > 1:
+                        raise IndexError("{}-dimensional boolean indexing is "
+                                         "not supported. ".format(variable.ndim))
+                    (variable,) = variable._nonzero()
+
                 variables.append(variable)
+                out_dims_set.update(variable.dims)
+
+        variable_dims = set()
+        for variable in variables:
+            variable_dims.update(variable.dims)
+
+        slices = []
+        for i, (dim, value) in enumerate(zip(self.dims, key)):
+            if isinstance(value, slice):
+                if dim in variable_dims:
+                    # We only convert slice objects to variables if they share
+                    # a dimension with at least one other variable. Otherwise,
+                    # we can equivalently leave them as slices and transpose the
+                    # result. This is significantly faster/more efficient for
+                    # most array backends.
+                    values = np.arange(*value.indices(self.sizes[dim]))
+                    variables.insert(i, Variable((dim,), values))
```

Comment:
I am wondering if this line is correct.
```python
v = Variable(['w', 'x', 'y', 'z'], [[[[1, 2, 3], [4, 5, 6]]]])
ind = Variable(['y'], [0])
v_new = v[ind, :, 1:2, 2]
```
raises
'IndexError: index 2 is out of bounds for axis 2 with size 2'